src/image/PhotoCalib.cc

```diff
@@ -0,0 +1,409 @@
+    }
+
+    PersistenceHelper(table::Schema const & s) :
+        schema(s),
+        countsMag0(s["countsMag0"]),
+        countsMag0Sigma(s["countsMag0Sigma"]),
+        isConstant(s["isConstant"]),
+        field(s["field"])
+    {}
+
+};
+
+class PhotoCalibFactory : public table::io::PersistableFactory {
+public:
+
+    virtual PTR(table::io::Persistable)
+    read(InputArchive const & archive, CatalogVector const & catalogs) const {
+        table::BaseRecord const & record = catalogs.front().front();
+        PersistenceHelper const & keys(record.getSchema());
+        return std::make_shared<PhotoCalib>(record.get(keys.countsMag0), record.get(keys.countsMag0Sigma),
+                                            archive.get<afw::math::BoundedField>(record.get(keys.field)),
+                                            record.get(keys.isConstant));
+    }
+
+    PhotoCalibFactory(std::string const & name) : afw::table::io::PersistableFactory(name) {}
+};
+
+std::string getPhotoCalibPersistenceName() { return "PhotoCalib"; }
+
+PhotoCalibFactory registration(getPhotoCalibPersistenceName());
+
+} // anonymous namespace
+
+std::string PhotoCalib::getPersistenceName() const { return getPhotoCalibPersistenceName(); }
+
+void PhotoCalib::write(OutputArchiveHandle & handle) const {
+    PersistenceHelper const & keys = PersistenceHelper::get();
+    table::BaseCatalog catalog = handle.makeCatalog(keys.schema);
+    auto record = catalog.addNew();
+    record->set(keys.countsMag0, _countsMag0);
+    record->set(keys.countsMag0Sigma, _countsMag0Sigma);
+    record->set(keys.isConstant, _isConstant);
+    record->set(keys.field, handle.put(_zeroPoint));
+    handle.saveCatalog(catalog);
+}
+
+// ------------------- private/protected helpers -------------------
+
+void PhotoCalib::countsToMaggiesArray(const afw::table::SourceCatalog &sourceCatalog,
+                                      const std::string &countsField,
+                                      ndarray::Array<double,2,2> result,
+                                      bool throwOnNegativeFlux) const {
+    double counts, countsSigma, maggies, countsMag0;
+    double negatives = 0;
+    auto countsKey = sourceCatalog.getSchema().find<double>(countsField+"_counts").key;
+    auto countsSigmaKey = sourceCatalog.getSchema().find<double>(countsField+"_countsSigma").key;
+    auto iter = result.begin();
+    for (auto const & rec: sourceCatalog) {
+        counts = rec.get(countsKey);
+        if (counts < 0)
+            negatives++;
+        countsSigma = rec.get(countsSigmaKey);
+        if (_isConstant)
+            countsMag0 = _countsMag0;
+        else
+            countsMag0 = _zeroPoint->evaluate(rec.getCentroid());
+        maggies = toMaggies(counts, countsMag0);
+        (*iter)[0] = maggies;
+        (*iter)[1] = toMaggiesSigma(counts, countsSigma, countsMag0, _countsMag0Sigma, maggies);
+        iter++;
+    }
+    if (negatives > 0  && throwOnNegativeFlux) {
```

Comment:
Per our discussion, I've completely removed the `throwOnNegative` functionality, thus mooting this question.